sympy/polys/euclidtools.py

```diff
@@ -1248,6 +1248,579 @@ def dup_zz_heu_gcd(f, g, K):
+    >>> x, y = symbols('x y')
+    >>> K = ZZ[x, y]
+    >>> p =[K(x**2 - y**2), K(x - y)]
+    >>> sparse_gcd(p)
+    x - y
+
+    """
+    ring = p[0].ring
+    domain = ring.domain
+
+    if any(len(pi) == 1 for pi in p):
+        return _gcd_terms(p, ring, domain)
+
+    p, monomial_gcd = gcd_list_monom(p)
+
+    p, common_symbols = _gcd_coeffs(p)
+
+    gcd = p[0]
+    for pi in p[1:]:
+        gcd = sparse_prs_gcd(gcd, pi)
+
+    if monomial_gcd is not None:
+        gcd = gcd * monomial_gcd
+
+    return gcd
+
+
+def _gcd_terms(p, ring, domain):
+    """
+    Computes the greatest common divisor (GCD) of a list of polynomials p in a given ring and domain.
+
+    Parameters
+    ==========
+        p (list): List of polynomials.
+        ring (PolynomialRing): Ring in which the polynomials are defined.
+        domain (Domain): Domain in which the coefficients are defined.
+
+    Returns
+    ==========
+        term_gcd (Polynomial): Polynomial representing the GCD of the input polynomials.
+
+    Examples
+    ==========
+    >>> from sympy import ZZ, symbols, ring, domain
+    >>> x, y = symbols('x, y')
+    >>> K = ZZ[x, y]
+    >>> p = [K(x**2 - y**2), K(x - y)]
+    >>> ring = p[0].ring
+    >>> domain = ring.domain
+    >>> _gcd_terms(p, ring, domain)
+    1
+
+    """
+    monomials = set()
+    coefficients = set()
+
+    for pi in p:
+        for monomial, coefficient in pi.terms():
+            monomials.add(monomial)
+            coefficients.add(coefficient)
+
+    monomial_gcd = gcd_monom(monomials)
+    coefficient_gcd = _ground_gcd(coefficients, domain)
+    term_gcd = ring({monomial_gcd: coefficient_gcd})
+
+    return term_gcd
+
+
+
+def _ground_gcd(coefficients, domain):
+    """
+    Computes the greatest common divisor (GCD) of multiple polynomials represented by their terms.
+
+    Parameters
+    ==========
+    coefficients (iterable): A list of coefficients.
+    domain: The domain in which the coefficients are defined.
+
+    Returns
+    ==========
+    The GCD of the coefficients.
+
+    Examples
+    ==========
+    >>> from sympy import ZZ
+    >>> coefficients = [3, 12, 6]
+    >>> _ground_gcd(coefficients, ZZ)
+    3
+
+    """
+
+    coefficients = list(coefficients)
+    gcd = domain.gcd
+    d = coefficients[0]
+
+    for coefficient in coefficients[1:]:
+        d = gcd(d, coefficient)
+
+        if d == domain.one:
+            break
+
+    return d
+
+
+def gcd_monom(monomials):
+    """
+    Computes the greatest common divisor (GCD) of the exponents for each variable in the monomials.
+    """
+
+    return monomial_gcd(*monomials)
+
+
+
+def gcd_list_monom(p):
+    """
+    Computes the greatest common divisor (GCD) of a list of polynomials p with respect to the monomials.
+    """
+    ring = p[0].ring
+    zero_monom = ring.zero_monom
+    monomials = chain(*p)
+    monomial_gcd = tuple(map(min, zip(*monomials)))
+
+    if monomial_gcd == ring.zero_monom:
+        return p, None
```

Comment:
A common case is that there is a large number of terms and the monomial gcd is just the zero monomial. We can discover that quickly without needing to loop over every monomial:

- If the zero monomial is present in the polynomial then we know immediately that we can return.
- Otherwise after looping through a few monomials we would find that the gcd was the zero monomial quite quickly and could exit the loop early.

Reply:
Pure monomial functions should be added in monomials.py which actually already has a function for this:
https://github.com/sympy/sympy/blob/68a011f75ec8a2c2edb2f3ee7de2349af0b0d567/sympy/polys/monomials.py#L258